File: Longest_Substring_Without_Repeating_Characters/Longest_Substring_Without_Repeating_Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<char, int> Map;
        int len=0;
        int rec=0;
        int s_len=s.length();
        for(int i=0;i<s_len;i++){
            if(Map.find(s[i])==Map.end()){
                Map[s[i]] = i;
            }else{
                len = max(len, i-rec);
                if(Map[s[i]] >= rec){
                    rec = Map[s[i]] + 1;
                }
                Map[s[i]] = i;
            }
        }
        len = max(len, s_len-rec);
        return len;
    }
};

File: Longest_Substring_Without_Repeating_Characters/Longest_Substring_Without_Repeating_Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        res = {}
        ans = 0
        left = 0
        for i in range(len(s)):
            if s[i] in res:
                left = max(left, res[s[i]]+1)
            ans = max(ans, i-left+1)
            res[s[i]] = i
        return ans;
